tutorial/03-configuration.md
# Code configuration

## Configuration sources

Run configuration refers to selecting things like initial conditions, the
physics to be modeled, the type of hardware (CPU or GPU), the accuracy of
the numerical integration scheme, and more. Sailfish uses a single data
structure, the `Sailfish` class, to specify a configuration. Each time you run
the code, a `Sailfish` class instance is created by gathering configuration
data from a sequence of inputs. Here is the procedure used by the code:

1. A default configuration is created. This default should not change as the code evolves.
2. The configuration is updated with key-value pairs from a _sequence_ of the
   following kinds of input:
	1. A JSON-formatted configuration file (ends in .json)
	2. A checkpoint file (ends in .pk)
	3. The name of a registered _preset_ (has no extension)
3. The configuration is finally updated with key-value pairs taken from command
   line flags

Configuration items coming later in the procedure take precedence. This
enables the creation of base configurations, which can be "tweaked" with input
from JSON files or from the command line.

A most-basic example is to start the code from a single JSON file. Try saving
the following code in a file called `my-run.json`,

```json
{
    "initial_data.model": "sod",
    "domain.num_zones": [200, 1, 1],
    "driver.tfinal": 0.1
}
```

and then run sailfish with that file supplied as input:

```bash
sailfish run my-run.json
```

This configuration is very close to the default one, so not many configuration
items need to be overridden. This is also a very easy and common test case, and
as such it is a Sailfish built-in _preset_. A preset is a function that is
built-in to the code and which emits these configuration items. This
particular configuration can be found in a function called `sod` in the
`sailfish.models` module, which looks like this:

```python
@preset
def sod():
    return {
        "initial_data.model": "sod",
        "domain.num_zones": [200, 1, 1],
        "driver.tfinal": 0.1,
    }
```

The decorator `@preset` tells the application that this function should be
recognized as a preset. Running `sailfish doc presets` prints a list of the
available presets. To invoke this preset, you would type on the command line
`sailfish run sod`.

Now suppose that you want to reuse _most_ of a particular preset, or
configuration file, but you also want to override certain pieces of it. You
have two options: (a) you can _chain_ the configuration files or (b) use
command line flags. Option (a) is more general, since there is not a command
line flag to control all the configuration items. For example if you wanted to
run the Sod problem but with higher resolution, you could put the following
text in a second JSON file `res.json`

```json
{
    "domain.num_zones": [20000, 1, 1],
}
```

and then invoke the code like this: `sailfish run sod res.json`. The
`domain.num_zones` configuration item from the JSON file takes precedence, so
you would get all the configuration items from the `sod` preset, except with
20,000 zones instead of 200. Since the grid resolution is a commonly modified
parameter, so there also is a command line flag `--resolution/-n` to control
it directly, which is generally more convenient:


```bash
sailfish run sod -n 20000
```

## The `Sailfish` class

You can read the full description of the code configuration in the
`sailfish.config` module. That module defines the `Sailfish` class, which
looks like this:

```python
class Sailfish:
    """
    Top-level application configuration struct

    Fields
    ------

    name:               a name given to the run
    driver:             runs the simulation main loop, handles IO
    physics:            the physics equations to be solved
    initial_data:       initial condition, used if not a restart
    boundary_condition: boundary condition to apply at domain edges
    domain:             the physical domain of the problem
    strategy:           the solution strategy; does not affect the numerical solution
    scheme:             algorithmic choices which can affect the solution accuracy
    forcing:            driving of the hydrodynamic fields to implement a soft BC
    """

    name: str = None
    driver: Driver = Driver()
    physics: Physics = Physics()
    initial_data: dict = None
    boundary_condition: BoundaryCondition = BoundaryCondition()
    domain: CoordinateBox = CoordinateBox()
    coordinates: Coordinates = "cartesian"
    strategy: Strategy = Strategy()
    scheme: Scheme = Scheme()
    forcing: Forcing = None
```

Several of the class attributes refer to configuration sub-classes, for
example `Driver`, `Physics`, and `Strategy`. You can also find these classes
in the `config` module. You can figure out everything about the possible code
configurations by reading the source code of that module.

Configuration data you supply through JSON files, presets, checkpoints, or
command line flags is first packed into a Python dictionary, and then
converted to a `Sailfish` struct and validated. The mapping of a Python
dictionary to the `Sailfish` class is pretty self-explanatory. It is allowable
to use either dot-syntax in the dictionary keys, or nested dictionaries, i.e.
the following two JSON configuration files are equivalent:

_Using nested dictionaries_
```json
{
    "initial_data": {
    	"model": "sod"
    }
}
```

_Using dot-syntax_
```json
{
    "initial_data.model": "sod"
}
```

 

## Model data classes


The `Sailfish.initial_data` attribute contains configuration items that will
be used to construct a class that provides initial conditions for the
hydrodynamic fields. When a new simulation is started, the solver will
construct initial data by calling a member function on the initial data class
with the signature `primitive(self, box: CoordinateBox) -> ndarray`. The
classes that provide initial data classes are also called "model data"
classes, and they are marked with the `@modeldata` decorator. This decorator
registers model data classes and allows configuration files to refer to those
classes by name (note that the name is converted from `TitleCase` to
`dash-case`).

The `sailfish.models` module contains many examples of model data classes. A
simple example is one that defines the initial conditions for a sinusoidal
density wave in a one-dimensional hydrodynamics problem:

```python
@modeldata
class DensityWave:
    """
    Sinusoidal density wave translating rigidly
    """

    amplitude: float = 0.2

    @property
    def primitive_fields(self):
        return "density", "x-velocity", "pressure"

    def primitive(self, box: CoordinateBox):
        x = box.cell_centers()
        p = zeros(x.shape + (3,))
        p[..., 0] = 1.0 + self.amplitude * sin(2 * pi * x)
        p[..., 1] = 1.0
        p[..., 2] = 1.0
        return p
```

This model data class defines one _model parameter_ called `amplitude`, which
can be controlled through the configuration data, for example:

```json
{
    "initial_data": {
        "model": "density-wave",
        "amplitude": 0.5
    }
}
```

Also note there is a (property) function called `primitive_fields`. This
property must be implemented by the model data class to return the names of
the primitive variable fields. This provides some self-documentation, but it
is also used by the solver to infer the _number_ of hydrodynamic fields that
will be evolved, which could be 3, 4, or 5 depending on the number of vector
components of the velocity field need to be evolved. Note that the ordering of
these fields is imposed by the solver, and is e.g. `[density, vx, pressure]`
if one vector component of the velocity field is evolved, or e.g. `[density,
vx, vy, vz, pressure]` if three vector components are evolved. Also note that
while the solver reads the _number_ of fields that are returned, it does not
actually read the names of the fields. The field names are there only for
self-documentation purposes. Especially with the velocity components, it could
make sense to name them e.g. `"density, x-velocity, pressure"` if your setup
is intended for a one-dimensional planar Cartesian geometry, or e.g.
`"density, radial-velocity", "polar-velocity, pressure"` if your setup is
intended for a two-dimensional axisymmetric spherical-polar coordinate system.


_sources/_autosummary/sailfish.system.configure_build.rst.txt
sailfish.system.configure\_build
================================

.. currentmodule:: sailfish.system

.. autofunction:: configure_build

sailfish/config.py
from math import pi
from typing import Literal, Union
from pydantic import Field
from .schema import schema
from .geometry import CoordinateBox


@schema
class Report:
    cadence: int = 100


@schema
class Checkpoint:
    cadence: float = 1.0
    format: Literal["none", "pickle"] = "pickle"


@schema
class Diagnostic:
    name: str


@schema
class Timeseries:
    cadence: float = 1.0
    diagnostics: list[Diagnostic] = None


@schema
class Driver:
    """
    Deals with the simulation control flow and input/output

    The simulation driver takes care of setting problem initial data,
    constructing solver instances, recording timeseries data (on-the-fly
    science products), and writing checkpoint files.

    Fields
    ------

    tstart: time when the simulation starts (does not need to be t=0)
    tfinal: time when the simulation ends
    report: number of iterations between one-line report messages
    timeseries: high-cadence recording of science products
    cfl_number: safety factor in the Courant-Friedrichs-Lewy condition
    new_timestep_cadence: iterations between recomputing the Courant-limited timestep
    """

    tstart: float = 0.0
    tfinal: float = 1.0
    report: Report = Report()
    checkpoint: Checkpoint = Checkpoint()
    timeseries: Timeseries = Timeseries()
    cfl_number: float = 0.1
    new_timestep_cadence: int = 1


@schema
class GammaLawEOS:
    """
    Adiabatic EOS with given gamma-law index
    """

    gamma_law_index: float = 5.0 / 3.0


@schema
class IsothermalEOS:
    """
    Isothermal EOS with a uniform sound speed
    """

    sound_speed: float = 1.0


@schema
class LocallyIsothermalEOS:
    """
    Isothermal EOS with a uniform nominal Mach number
    """

    mach_number: float = 10.0


EquationOfState = Union[GammaLawEOS, IsothermalEOS, LocallyIsothermalEOS]


@schema
class ConstantNu:
    nu: float


@schema
class ConstantAlpha:
    alpha: float


ViscosityModel = Union[ConstantNu, ConstantAlpha]


@schema
class Physics:
    """
    Physics equations to be solved

    Fields
    ------

    equation_of_state: thermodynamic EOS
    metric:            currently either newtonian or minkowski
    """

    equation_of_state: EquationOfState = GammaLawEOS()
    metric: Literal["newtonian", "minkowski"] = "newtonian"
    viscosity: ViscosityModel = None


Reconstruction = Union[Literal["pcm"], tuple[Literal["plm"], float]]
TimeIntegration = Literal["fwd", "rk1", "rk2", "rk3"]


@schema
class Scheme:
    """
    Algorithm parameters for the solution scheme

    A scheme refers to the parts of a solver that affect the numerical
    solution up to the truncation order.

    Fields
    ------

    reconstruction:    spatial reconstruction method
    time_integration:  time-integration scheme (fwd|rk1|rk2|rk3)
    """

    reconstruction: Reconstruction = "pcm"
    time_integration: TimeIntegration = "fwd"


@schema
class Strategy:
    """
    Algorithm parameters only affecting performance

    With everything else held fixed, differences in the solver strategy should
    only affect how fast the code runs, and not the numerical solution it
    generates.

    Fields
    ------

    hardware:    compute device [cpu|gpu]
    data_layout: array-of-struct (fields-last) or struct-of-array (fields-first)
    cache_flux:  one Riemann problem per face, difference the resulting array
    cache_prim:  pre-compute primitive quantities, vs. re-compute over stencil
    num_patches: decompose domain to enable threads, streams, or multiple GPU's
    num_threads: use a thread pool of this size to drive a multi-patch solver
    gpu_streams: use the per-thread-default-stream, or one stream per grid patch
    """

    hardware: Literal["cpu", "gpu"] = "cpu"
    data_layout: Literal["fields-first", "fields-last"] = "fields-last"
    cache_flux: bool = False
    cache_prim: bool = False
    cache_grad: bool = False
    num_patches: int = 1
    num_threads: int = 1
    gpu_streams: Literal["per-thread", "per-patch"] = "per-thread"

    @property
    def transpose(self):
        """
        synonym for fields-first (also called struct-of-arrays) data layout
        """
        return self.data_layout == "fields-first"


BoundaryConditionType = Literal["outflow", "periodic", "reflecting"]
Coordinates = Literal["cartesian", "spherical-polar", "cylindrical-polar"]


@schema
class BoundaryCondition:
    lower_i: BoundaryConditionType = None
    upper_i: BoundaryConditionType = None
    lower_j: BoundaryConditionType = None
    upper_j: BoundaryConditionType = None
    lower_k: BoundaryConditionType = None
    upper_k: BoundaryConditionType = None


@schema
class Forcing:
    """
    Soft boundary condition, drives fields to some value in a defined region

    rate:   rate at which the solution is driven to the target value
    ramp:   distance over which the region
    where:  a string describing the zone in which the buffer is active
    """

    rate: float = 1.0
    ramp: float = 0.0
    where: str = "x < 0.0"
    target: Literal["initial-data"] = "initial-data"

    def rate_array(self, box: CoordinateBox):
        """
        Return an array of rates at the zones centers of the given box
        """
        from math import e

        coordinate, inequality, value = self.where.split()
        s0 = float(value)
        x, y, z = box.cell_centers(dim=3)
        R = (x**2 + y**2) ** 0.5
        s = dict(x=x, y=y, z=z, R=R)[coordinate]
        sign = {"<": -1.0, ">": +1.0}[inequality]

        if self.ramp == 0.0:
            step = lambda x: x > 0.0
        else:
            step = lambda x: 1.0 / (1.0 + e ** (-x / self.ramp))

        return self.rate * step(sign * (s - s0))


@schema
class Sailfish:
    """
    Top-level application configuration struct

    Fields
    ------

    name:               a name given to the run
    driver:             runs the simulation main loop, handles IO
    physics:            the physics equations to be solved
    initial_data:       initial condition, used if not a restart
    boundary_condition: boundary condition to apply at domain edges
    domain:             the physical domain of the problem
    strategy:           the solution strategy; does not affect the numerical solution
    scheme:             algorithmic choices which can affect the solution accuracy
    forcing:            driving of the hydrodynamic fields to implement a soft BC
    """

    name: str = None
    driver: Driver = Driver()
    physics: Physics = Physics()
    initial_data: dict = None
    boundary_condition: BoundaryCondition = BoundaryCondition()
    domain: CoordinateBox = CoordinateBox()
    coordinates: Coordinates = "cartesian"
    strategy: Strategy = Strategy()
    scheme: Scheme = Scheme()
    forcing: Forcing = None

    def initialize(self):
        from dataclasses import asdict
        from .models import get_model_data_class

        # The model data class instance is computed after the constructor,
        # from the `initial_data` dictionary. Then the dictionary is replaced
        # with the model data class instance. The reason is that all of the
        # model data classes which compose the type union cannot be known at
        # the time the Sailfish class is first defined. Otherwise it imposes
        # confusing constraints on the order in which this module is imported,
        # relative to user modules that could create new model data classes.
        #
        model_cls = get_model_data_class(self.initial_data["model"])
        self.initial_data = model_cls(**self.initial_data)
        self.initial_data.physics = self.physics
        self.initial_data.coordinates = self.coordinates
        self.initial_data.boundary_condition = self.boundary_condition

        # Internal solver functions assume that all the fleshed-out axes come
        # at the beginning of the domain shape. Generalizing this was too much
        # of a headache, and the constraint should not be a problem from the
        # user perspective.
        #
        if self.domain.num_zones.index(1) < self.domain.dimensionality:
            raise ValueError(
                f"domain.num_zones = {self.domain.num_zones}; "
                f"squeezed axes must be at the end"
            )

        # The viscosity implementation is limited to 2d cartesian coordinates.
        # It also requires primitive data and gradients to be cached.
        #
        if type(self.physics.viscosity) is ConstantNu:
            if self.coordinates != "cartesian":
                raise ValueError("viscosity only implemented in cartesian coordinates")
            if self.domain.dimensionality != 2:
                raise ValueError("viscosity only implemented in 2d")
            if self.domain.grid_spacing[0] != self.domain.grid_spacing[1]:
                raise ValueError("viscosity requires isotropic grid spacing (dx = dy)")
            if not self.strategy.cache_prim:
                raise ValueError("viscosity requires strategy.cache_prim")
            if not self.strategy.cache_grad:
                raise ValueError("viscosity requires strategy.cache_grad")

        # Constant-alpha viscosity is not yet implemented.
        #
        if type(self.physics.viscosity) is ConstantAlpha:
            raise ValueError("constant-alpha viscosity is not implemented yet")


def parse_num_zones(arg):
    """
    Promote an integer or two-tuple to a three-tuple of integers

    This factory function is used by the argparse type parameter to convert
    user input to a domain.num_zones parameter.
    """
    res = tuple(int(i) for i in arg.split(","))

    if len(res) == 1:
        return res + (1, 1)
    if len(res) == 2:
        return res + (1,)
    if len(res) == 3:
        return res
    raise ValueError(f"invalid argument for num_zones {arg}")


def parse_reconstruction(arg):
    """
    Promote a string to a reconstruction model

    This factory function is used by the argparse type parameter to convert
    user input to a scheme.reconstruction parameter.
    """
    try:
        mode, theta = arg.split(":")
        return mode, float(theta)
    except ValueError:
        if arg == "plm":
            return arg, 1.5
        else:
            return arg


def parse_checkpoint(arg):
    """
    Promote a string to a checkpoint model

    This factory function is used by the argparse type parameter to convert
    user input to a driver.checkpoint parameter.
    """
    if ":" in arg:
        return Checkpoint(*arg.split(":"))
    else:
        return Checkpoint(arg)


def add_config_arguments(parser: "argparser.ArgumentParser"):
    """
    Add arguments to a parser controlling a subset of a Sailfish config struct
    """

    parser.add_argument(
        "--mode",
        "--hardware",
        dest="strategy.hardware",
        choices=Strategy.type_args("hardware"),
        help="execution mode",
    )
    parser.add_argument(
        "-n",
        "--num-zones",
        "--resolution",
        type=parse_num_zones,
        dest="domain.num_zones",
        metavar="N",
    )
    parser.add_argument(
        "--patches",
        type=int,
        default=None,
        metavar="P",
        help=Strategy.describe("num_patches"),
        dest="strategy.num_patches",
    )
    parser.add_argument(
        "--threads",
        type=int,
        default=None,
        metavar="T",
        help=Strategy.describe("num_threads"),
        dest="strategy.num_threads",
    )
    parser.add_argument(
        "--streams",
        type=str,
        choices=Strategy.type_args("gpu_streams"),
        metavar="S",
        help=Strategy.describe("gpu_streams"),
        dest="strategy.gpu_streams",
    )
    parser.add_argument(
        "--cfl",
        "--cfl-number",
        metavar="C",
        help=Driver.describe("cfl_number"),
        dest="driver.cfl_number",
    )
    parser.add_argument(
        "--new-timestep-cadence",
        help=Driver.describe("new_timestep_cadence"),
        metavar="N",
        dest="driver.new_timestep_cadence",
    )
    parser.add_argument(
        "-m",
        "--time-integration",
        choices=Scheme.type_args("time_integration"),
        help=Scheme.describe("time_integration"),
        dest="scheme.time_integration",
    )
    parser.add_argument(
        "-r",
        "--reconstruction",
        type=parse_reconstruction,
        help=Scheme.describe("reconstruction"),
        dest="scheme.reconstruction",
        metavar="R:P",
    )
    parser.add_argument(
        "-e",
        "--tfinal",
        type=float,
        help=Driver.describe("tfinal"),
        dest="driver.tfinal",
        metavar="T",
    )
    parser.add_argument(
        "-f",
        "--fold",
        "--report-cadence",
        type=int,
        help=Report.describe("cadence"),
        dest="driver.report.cadence",
        metavar="F",
    )
    parser.add_argument(
        "--checkpoint",
        "-c",
        type=parse_checkpoint,
        dest="driver.checkpoint",
        metavar="C:F",
    )
    parser.add_argument(
        "--timeseries",
        "-t",
        type=float,
        dest="driver.timeseries.cadence",
        metavar="T",
    )
    parser.add_argument(
        "--data-layout",
        type=str,
        choices=Strategy.type_args("data_layout"),
        help=Strategy.describe("data_layout"),
        dest="strategy.data_layout",
    )
    parser.add_argument(
        "--cache-prim",
        action="store_true",
        dest="strategy.cache_prim",
        default=None,
    )
    parser.add_argument(
        "--cache-flux",
        action="store_true",
        dest="strategy.cache_flux",
        default=None,
    )
    parser.add_argument(
        "--cache-grad",
        action="store_true",
        dest="strategy.cache_grad",
        default=None,
    )
    parser.add_argument(
        "--metric",
        type=str,
        choices=Physics.type_args("metric"),
        help=Physics.describe("metric"),
        dest="physics.metric",
    )


if __name__ == "__main__":
    from rich.console import Console
    from rich.live import Live

    console = Console()
    config = Sailfish(physics=Physics(equation_of_state=IsothermalEOS()))

    for field in config.__dataclass_fields__:
        value = getattr(config, field)
        if hasattr(value, "__configmodel__"):
            console.print()
            console.print(value)
